### ISSUE TITLE
Generate dependent items on arbitrary solution files with one master csx file. eg SQL code wrapper

### ISSUE THREAD

**Issue by williamdenton** [daveaglick/Scripty#22]
I currently use [T4](https://msdn.microsoft.com/en-us/library/bb126445.aspx) to generate wrapper classes to execute sql with [dapper](https://github.com/StackExchange/dapper-dot-net). To do this I use [T4Toolbox](https://github.com/olegsych/T4Toolbox). 
This works ~~well~~ _ok_ but T4 has some drawbacks:
- native VS support is non existent (resharper helps)
- generated code formatting is hard to get right (T4 escapes code and renders everything else as output vs csx is code and explicitly renders output)
- spaghetti code
- unreadable and hard to follow
- it's not Scripty

It would be great to attach a custom tool to my `*.sql` files which executes a script file to build a wrapper class that allows me to easily execute my sql

To do this I would like to:
- Set the CustomTool property in the file project properties window to "ScriptyGenerator"
- The Visual Studio Extension adds additional field to Properties window to specify path to csx file
- csx file is automatically executed when ever referenced file is saved
- generated code is attached as nested item on referenced file

Desired project structure:
![image](https://cloud.githubusercontent.com/assets/7498040/16252507/e64031b4-3882-11e6-996a-fe2faf0b74a8.png)
where the file name `Sample.generated.cs` is under control of the script 

Desired file properties window:
![image](https://cloud.githubusercontent.com/assets/7498040/16252663/3c910f38-3884-11e6-91d5-49ba3ff7812f.png)

This shouldn't be limited to `sql` files as i can see this being useful for `xml`, `json` and many other file types where you would want to generate code based on some file.

**Comment by daveaglick** [comment]
> it's not Scripty

Lol.

I'm not sure how technically possible this request is. Or I guess I should say how feasible - from other VS addins I've seen, just about anything is possible but the question is usually how much effort you're willing to dedicate. Setting the Custom Tool property kicks off what is known as a _Single File Generator_. The SFG is basically a special class that does a bunch of COM interop. It receives the text content of the file that uses it as a Custom Tool. To get it to read the alternate "Custom Tool Template" (or whatever this property would be called), we would need to add that new property to VS, recognize it in the SFG, and hot-swap the content before the processing begins.

That said, I can see why this would be valuable. Particularly with regards to "csx file is automatically executed when ever referenced file is saved." Since the Custom Tool property only triggers processing when it's own file is saved, unless you can set the custom tool on a different input file, it won't trigger processing.

What about this alternate design: perhaps if we set the Custom Tool property for a non-csx file we can look for a corresponding `.csx` with the same name. If one is found, we can hot-swap the content inside the SFG before processing. This would be easier because there'd be no additional property to manage. It does mean the combination of extra file to csx would have to be one-to-one and by strict naming convention. Would you want to use the same `.csx` for more than one `.sql` file (I'm guessing yes)?

Let me think on this. However. even if it looks like it's doable it probably won't be soon since it'll take some work and there is other low-hanging fruit I'd like to get to first. But it's a great idea.

**Comment by salmelo** [comment]
May want to look into https://marketplace.visualstudio.com/items?itemName=ThomasLevesque.AutoRunCustomTool .
Don't know if it's worth having Scripty try to do all that itself.

**Comment by CZEMacLeod** [comment]
I was about to suggest that myself. I use the AutoRunCustomTool just now to do this for several tt templates, although I use scripty for the more complex stuff and just do it with MSBuild on build without any custom tool.
